src/streamed/index.ts

```diff
@@ -133,16 +176,77 @@ export class StreamedOrderbook {
         fromBlock + this.options.blockPageSize - 1,
         endBlock,
       )
-      this.options.logger?.debug(`fetching page ${fromBlock}-${toBlock}`)
 
+      this.options.logger?.debug(`fetching page ${fromBlock}-${toBlock}`)
       const events = await this.contract.getPastEvents(
         "allEvents",
         { fromBlock, toBlock },
       )
 
+      this.options.logger?.debug(`applying ${events.length} past events`)
       this.state.applyEvents(events)
     }
   }
+
+  /**
+   * Apply new confirmed events to the account state and store the remaining
+   * events that are subject to reorgs into the `pendingEvents` array.
+   *
+   * @remarks
+   * If there is an error retrieving the latest events from the node, then the
+   * account state remains unmodified. This allows the updating orderbook to be
+   * more fault-tolerant and deal with nodes being temporarily down and some
+   * intermittent errors.
+   *
+   * @throws
+   * Throws if there was an error applying the newly confirmed events.
+   */
+  private async applyNewEvents(): Promise<void> {
+    const fromBlock = this.state.nextBlock
+
+    let confirmedEvents: EventData[]
+    let pendingEvents: EventData[]
+    try {
+      this.options.logger?.debug(`fetching new events from ${fromBlock}-latest`)
+      const events = await this.contract.getPastEvents("allEvents", { fromBlock })
+
+      const latestBlock = await this.web3.eth.getBlockNumber()
+      const confirmedBlock = latestBlock - this.options.blockConfirmations
+      const confirmedEventCount = events.findIndex(ev => ev.blockNumber <= confirmedBlock)
+
+      confirmedEvents = events.splice(0, confirmedEventCount)
+      pendingEvents = events
+    } catch (err) {
+      this.options.logger?.warn(`error retrieving new events: ${err}`)
+      return
+    }
+
+    if (confirmedEvents.length > 0) {
+      this.options.logger?.debug(`applying ${confirmedEvents.length} confirmed events`)
+      this.state.applyEvents(confirmedEvents)
+    }
+    this.pendingEvents = pendingEvents
+  }
+
+  /**
+   * Schedule an orderbook update.
+   */
+  private scheduleUpdate(): void {
+    this.updateTimeout = setTimeout(
+      async () => {
+        try {
+          await this.applyNewEvents()
+          if (this.updateTimeout) {
+            this.scheduleUpdate()
+          }
+        } catch (err) {
+          this.options.logger?.error(`error applying new events: ${err}`)
+          this.updateError = err
```

Comment:
I read your comment on the last PR. I still don't fully understand how you intend to use it. Should we not store all error occurrences and some additional information (e.g. the block number) to make this error more useful?

Reply:
> Should we not store all error occurrences

These are unrecoverable errors. Errors during polling, for example, where we want to continue with the fiber just produce logging warnings.

> Should we not store all error occurrences and some additional information

We already log the block range at the beginning of each `applyNewEvents` but we can also include them in the error message to make debugging a bit easier.

Reply:
The idea behind this `updateError` is more to indicate to callers (like the request handler that tries to read the current orderbook from the `StreamedOrderbook` instance that it is currently in an error state and not producing a valid orderbook. So it returns the last unrecoverable error that it encountered when applying events.

Reply:
I see. The thing I was missing is that we don't continue after this is set. We do sometimes have transient errors, so another (not necessarily better) approach could be to schedule updates regardless of errors.